sdk/test/trace/batch_span_processor_test.cc

```diff
@@ -0,0 +1,344 @@
+}
+
+TEST_F(BatchSpanProcessorTestPeer, TestForceFlush)
+{
+    std::shared_ptr<bool> is_shutdown(new bool(false));
+    std::shared_ptr<std::vector<std::unique_ptr<sdk::trace::SpanData>>> spans_received(
+        new std::vector<std::unique_ptr<sdk::trace::SpanData>>);
+
+    auto batch_processor = GetMockProcessor(spans_received, is_shutdown);
+    const int num_spans = 3;
+
+    auto test_spans = GetTestSpans(batch_processor, num_spans);
+
+    for(int i = 0; i < num_spans; ++i){
+        batch_processor->OnEnd(std::move(test_spans->at(i)));
+    }
+
+    // Wait a bit here too for the same reasons as mentioned in the previous test (TestShutdown)
+    // Helps in proper preemption of the worker thread
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+
+    // force flush
+    batch_processor->ForceFlush(std::chrono::milliseconds(100));
+
+    EXPECT_EQ(num_spans, spans_received->size());
+    for(int i = 0; i < num_spans; ++i)
+    {
+        EXPECT_EQ("Span " + i, spans_received->at(i)->GetName());
+    } 
+
+    // Create some more spans to make sure that the processor still works
+    auto more_test_spans = GetTestSpans(batch_processor, num_spans);
+    for(int i = 0; i < num_spans; ++i){
+        batch_processor->OnEnd(std::move(more_test_spans->at(i)));
+    }
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    batch_processor->ForceFlush(std::chrono::milliseconds(100));
+
+    EXPECT_EQ(num_spans*2, spans_received->size());
+    for(int i = 0; i < num_spans; ++i)
+    {
+        EXPECT_EQ("Span " + i%3, spans_received->at(i)->GetName());
+    } 
+    // Wait a bit here so that the worker thread can go back to waiting on condition
+    // variable and the shutdown() method can subsequently wake the worker thread
+    // TODO: Discuss this
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+}
+
+// TODO : Discuss and revise this function (maybe force_flush can return a bool)
+TEST_F(BatchSpanProcessorTestPeer, TestForceFlushTimeout)
+{
+    std::shared_ptr<bool> is_shutdown(new bool(false));
+    std::shared_ptr<std::vector<std::unique_ptr<sdk::trace::SpanData>>> spans_received(
+        new std::vector<std::unique_ptr<sdk::trace::SpanData>>);
+
+    const std::chrono::milliseconds export_delay(300);
+    const int num_spans = 3;
+
+    auto batch_processor = GetMockProcessor(spans_received, is_shutdown, export_delay);
+
+    auto test_spans = GetTestSpans(batch_processor, num_spans);
+
+    for(int i = 0; i < num_spans; ++i){
+        batch_processor->OnEnd(std::move(test_spans->at(i)));
+    }
+
+    // Helps in proper preemption of the worker thread
+    // TODO: Discuss this
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+
+    // force flush
+    batch_processor->ForceFlush(std::chrono::milliseconds(100));
+
+    // NOTE: It is also possible that some but not all spans were exported
+    // TODO: Discuss this 
+    EXPECT_EQ(0, spans_received->size());
```

Comment:
It is hard to test whether the ForceFlush() method times out or not because it does not have a return type. Making it return a bool would help.

Reply:
If ForceFlush is called and is still in progress, would the processor take newly finished spans?

I've done some work in [OpenCensus Python](https://github.com/census-instrumentation/opencensus-python/blob/b7c7a67d7bcae749c19a3508777522f206da75cf/contrib/opencensus-ext-azure/opencensus/ext/azure/common/exporter.py#L84) to make this more deterministic, it was considered as complex and not a generally interesting to put as the core SDK.

Reply:
If ForceFlush is in progress I think it should first drain out whatever is in the queue at that very moment and continue.
This is exactly what happens in the [Java implementation](https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java#L249). 
I think its a good approach.

Thank you for the link! Correct me if I am wrong, but, is that for spurious wake-ups.
If so, I am planning on doing something the exact same thing for handling them.

Reply:
Sounds great to me!

Reply:
Thank you! 
I'll change both the processor and the spurious wake-up handling accordingly👍